File: internal/combo/combo.go
// Package combo resolves a logical `model` string into an ordered list of
// (provider, upstreamModel) attempts. Three shapes supported, matching the
// Node services/combo.js precisely:
//
//	combo:<slug>      → look up combos.nodes for the tenant
//	<provider>:<m>    → single attempt
//	<bare model>      → heuristic provider mapping
package combo

import (
	"context"
	"encoding/json"
	"regexp"
	"strings"

	"github.com/amaozhao/lazirouter/internal/db"
	"github.com/amaozhao/lazirouter/internal/errs"
)

// Attempt is one node in the fallback chain.
type Attempt struct {
	Provider      string
	UpstreamModel string
	ConnectionID  *int64
	Step          int
	SourceCombo   string
}

type comboNode struct {
	Provider     string `json:"provider"`
	Model        string `json:"model"`
	ConnectionID *int64 `json:"connection_id,omitempty"`
	Weight       int    `json:"weight,omitempty"`
}

var gptRe = regexp.MustCompile(`^(gpt-|o\d|chatgpt-)`)

// Resolve returns the attempts in order, or a ValidationError.
func Resolve(ctx context.Context, tenantID int64, modelInput string) ([]Attempt, error) {
	if modelInput == "" {
		return nil, errs.Validation("model must be a non-empty string")
	}

	if strings.HasPrefix(modelInput, "combo:") {
		slug := modelInput[6:]
		var nodesJSON []byte
		err := db.Pool().QueryRow(ctx, `
			SELECT nodes FROM combos
			WHERE tenant_id = $1 AND slug = $2 AND enabled = TRUE LIMIT 1
		`, tenantID, slug).Scan(&nodesJSON)
		if err != nil {
			return nil, errs.Validation("Unknown combo: "+slug,
				map[string]any{"hint": "create it via POST /api/combos"})
		}
		var nodes []comboNode
		if err := json.Unmarshal(nodesJSON, &nodes); err != nil || len(nodes) == 0 {
			return nil, errs.Validation("Combo " + slug + " has no nodes")
		}
		out := make([]Attempt, len(nodes))
		for i, n := range nodes {
			out[i] = Attempt{
				Provider:      strings.ToLower(n.Provider),
				UpstreamModel: n.Model,
				ConnectionID:  n.ConnectionID,
				Step:          i,
				SourceCombo:   slug,
			}
		}
		return out, nil
	}

	// explicit provider:model
	if idx := strings.Index(modelInput, ":"); idx > 0 && idx < len(modelInput)-1 {
		return []Attempt{{
			Provider:      strings.ToLower(modelInput[:idx]),
			UpstreamModel: modelInput[idx+1:],
			Step:          0,
		}}, nil
	}

	// heuristic — subscription-first defaults for claude/gpt/o-series
	provider := "openai"
	switch {
	case strings.HasPrefix(modelInput, "glm-"):
		provider = "glm"
	case strings.HasPrefix(modelInput, "deepseek-"):
		provider = "deepseek"
	case strings.HasPrefix(modelInput, "claude-"):
		provider = "claude"
	case strings.HasPrefix(modelInput, "gemini-"):
		provider = "gemini"
	case strings.HasPrefix(modelInput, "mock-"):
		provider = "mock"
	case gptRe.MatchString(modelInput):
		provider = "codex"
	}
	return []Attempt{{Provider: provider, UpstreamModel: modelInput, Step: 0}}, nil
}

File: cloud/router/src/services/combo.js
// Combo resolver — turns a logical model name into an ordered list of
// (provider, upstream_model) attempts. Two input shapes are supported:
//
//   1. `combo:<slug>` — look up combos.nodes for the tenant.
//      nodes = [{ provider, model, connection_id?, weight? }, ...]
//
//   2. `<provider>:<model>` — single-attempt resolution, no fallback.
//
//   3. `<bare-model>` — heuristic provider detection (kept for OpenAI compat).
//
// The result is an array of attempt descriptors; the route layer iterates
// them and stops at the first success.

import { query, ValidationError } from '@9router-cloud/shared';

export async function resolveAttempts(tenantId, modelInput) {
  if (!modelInput || typeof modelInput !== 'string') {
    throw new ValidationError('model must be a non-empty string');
  }

  if (modelInput.startsWith('combo:')) {
    const slug = modelInput.slice(6);
    const { rows } = await query(`
      SELECT nodes
      FROM combos
      WHERE tenant_id = $1 AND slug = $2 AND enabled = TRUE
      LIMIT 1
    `, [tenantId, slug]);
    if (rows.length === 0) {
      throw new ValidationError(`Unknown combo: ${slug}`,
        { hint: 'create it via POST /api/combos' });
    }
    const nodes = rows[0].nodes;
    if (!Array.isArray(nodes) || nodes.length === 0) {
      throw new ValidationError(`Combo ${slug} has no nodes`);
    }
    return nodes.map((n, i) => ({
      provider: String(n.provider).toLowerCase(),
      upstreamModel: String(n.model),
      connectionId: n.connection_id != null ? Number(n.connection_id) : null,
      step: i,
      sourceCombo: slug,
    }));
  }

  // Single attempt
  const idx = modelInput.indexOf(':');
  if (idx > 0 && idx < modelInput.length - 1) {
    return [{
      provider: modelInput.slice(0, idx).toLowerCase(),
      upstreamModel: modelInput.slice(idx + 1),
      connectionId: null,
      step: 0,
      sourceCombo: null,
    }];
  }

  // Heuristic for raw model id (matches the historical 9router behaviour).
  // GPT and o-series default to `codex` — the ChatGPT subscription path —
  // because that's the only OpenAI-side route a user normally has without
  // an API key. Callers who want the api-key OpenAI path can still send
  // `openai:gpt-4` explicitly.
  let provider = 'openai';
  if (modelInput.startsWith('glm-')) provider = 'glm';
  else if (modelInput.startsWith('deepseek-')) provider = 'deepseek';
  else if (modelInput.startsWith('claude-')) provider = 'anthropic';
  else if (modelInput.startsWith('gemini-')) provider = 'gemini';
  else if (modelInput.startsWith('mock-')) provider = 'mock';
  else if (/^(gpt-|o\d|chatgpt-)/i.test(modelInput)) provider = 'codex';

  return [{ provider, upstreamModel: modelInput, connectionId: null, step: 0, sourceCombo: null }];
}
